source/mesh_deformation/free_surface.cc

```diff
@@ -0,0 +1,417 @@
+                                                                    cell_dof_indices, mass_matrix, rhs, false);
+              }
+
+      rhs.compress (VectorOperation::add);
+      mass_matrix.compress(VectorOperation::add);
+
+      // Jacobi seems to be fine here.  Other preconditioners (ILU, IC) run into troubles
+      // because the matrix is mostly empty, since we don't touch internal vertices.
+      LinearAlgebra::PreconditionJacobi preconditioner_mass;
+      preconditioner_mass.initialize(mass_matrix);
+
+      SolverControl solver_control(5*rhs.size(), this->get_parameters().linear_stokes_solver_tolerance*rhs.l2_norm());
+      SolverCG<LinearAlgebra::Vector> cg(solver_control);
+      cg.solve (mass_matrix, dist_solution, rhs, preconditioner_mass);
+
+      mass_matrix_constraints.distribute (dist_solution);
+      output = dist_solution;
+    }
+
+
+    template <int dim>
+    void
+    FreeSurface<dim>::deformation_constraints(const DoFHandler<dim> &free_surface_dof_handler,
+                                              ConstraintMatrix &mesh_velocity_constraints) const
+    {
+      // For the free surface indicators we constrain the displacement to be v.n
+      LinearAlgebra::Vector boundary_velocity;
+
+      const IndexSet mesh_locally_owned = free_surface_dof_handler.locally_owned_dofs();
+      IndexSet mesh_locally_relevant;
+      DoFTools::extract_locally_relevant_dofs (free_surface_dof_handler,
+                                               mesh_locally_relevant);
+      boundary_velocity.reinit(mesh_locally_owned, mesh_locally_relevant, this->get_mpi_communicator());
+      project_velocity_onto_boundary(free_surface_dof_handler,mesh_locally_owned,mesh_locally_relevant,boundary_velocity);
+
+      // now insert the relevant part of the solution into the mesh constraints
+      IndexSet constrained_dofs;
+      DoFTools::extract_boundary_dofs(free_surface_dof_handler, ComponentMask(dim, true),
+                                      constrained_dofs, this->get_parameters().free_surface_boundary_indicators);
+      for (unsigned int i = 0; i < constrained_dofs.n_elements();  ++i)
+        {
+          types::global_dof_index index = constrained_dofs.nth_index_in_set(i);
+          if (mesh_velocity_constraints.can_store_line(index))
+            if (mesh_velocity_constraints.is_constrained(index)==false)
+              {
+                mesh_velocity_constraints.add_line(index);
+                mesh_velocity_constraints.set_inhomogeneity(index, boundary_velocity[index]);
+              }
+        }
+    }
+
+
+
+    template <int dim>
+    void FreeSurface<dim>::declare_parameters(ParameterHandler &prm)
+    {
+      prm.enter_subsection ("Mesh deformation");
+      {
+        prm.declare_entry("Free surface stabilization theta", "0.5",
+                          Patterns::Double(0,1),
+                          "Theta parameter described in Kaus et. al. 2010. "
```

Comment:
```suggestion
                          "Theta parameter described in Kaus et al. (2010). "
```
Extra credit if you add the complete reference to `doc/manual/manual.bib` and `\\cite{...}` it here

Reply:
It's already there :)